README.md
<p align="center">
  <br>
 <img width="500" src="https://user-images.githubusercontent.com/59066341/129483451-4196e1bb-f094-4b3c-aefc-41d77aff8117.png" alt="Package Logo">
 <br>
  <br>
 <a href="https://pub.dev/packages/field_suggestion">
   <img src="https://img.shields.io/badge/Special%20Made%20for-FieldSuggestion-blue" alt="License: MIT"/>
 </a>
 <a href="https://github.com/theiskaa/highlightable-text/blob/main/LICENSE">
   <img src="https://img.shields.io/badge/License-MIT-red.svg" alt="License: MIT"/>
 </a>
 <a href="https://github.com/theiskaa/highlightable-text/blob/main/CONTRIBUTING.md">
   <img src="https://img.shields.io/badge/Contributions-Welcome-brightgreen" alt="CONTRIBUTING"/>
 </a>

</p>

## Installing
- See the official installing guideline from [highlightable/install](https://pub.dev/packages/highlightable/install)

## Usage

### _Basic usage:_
```dart
HighlightText(
  'Only numbers: [1, 25, 50, ...] will be highlighted',
  // would highlight only numbers.
  highlight: const Highlight(pattern: r'\d'),
)
```

<img width="600" alt="basic-usage" src="https://user-images.githubusercontent.com/59066341/159167993-31854ab2-011f-4138-97ae-9c83fc202181.png">

### _Custom usage:_
```dart     
HighlightText(                                   
  "Hello, Flutter!",                             
  // Would highlight only the "Flutter" full word 'cause [detectWords] is enabled.  
  highlight: const Highlight(                          
    words: ["Flutter"],                          
  ),                                             
  caseSensitive: true, // Turn on case-sensitive.
  detectWords: true,  // Turn on full-word-detection.
  style: const TextStyle(                              
    fontSize: 25,                                
    color: Colors.black,                         
    fontWeight: FontWeight.bold,                 
  ),                                             
  highlightStyle: const TextStyle(                     
    fontSize: 25,                                
    letterSpacing: 2.5,                          
    color: Colors.white,                         
    backgroundColor: Colors.blue,                
    fontWeight: FontWeight.bold,                 
  ),                                             
)
```
<img width="400" alt="custom-usage" src="https://user-images.githubusercontent.com/59066341/159168147-a565e5a6-fcad-4f44-908b-e472ce1517f9.png">


CHANGELOG.md
## 1.0.5 - (22/03/22)

### Updates:

Resolved [#12](https://github.com/theiskaa/highlightable/issues/12) 
 - Updated state behaviour type of [HighlightText] widget.
 - Improved default-text-determining by theme.

## 1.0.4 - (20/03/22)

### Updates:

Resolved [#5](https://github.com/theiskaa/highlightable/issues/5) 
 - Re-structured the whole widget to improve rendering speed.
 - Added a new model object to pass Regular-Expression patterns, as higlighting search options.

**The new widget structure:**
```
 ╭──────╮        Highlight                                
 │ Data │       ╭─────────────────────────────────╮       
 ╰──────╯       │ ╭─────────╮   ╭───────────────╮ │       
    │       ╭──▶│ │ Pattern │ & │ Words/Letters │ │       
    │       │   │ ╰─────────╯   ╰───────────────╯ │       
    │       │   ╰─────────────────────────────────╯       
    ╰───────╯                                             
        │                                                 
    ╭── ▼ ──╮         ╭─────────────────────────────────╮ 
    │ Parser ───▶ ... │ Highlighted Data as Text Widget │ 
    ╰───────╯         ╰─────────────────────────────────╯ 
```

## 1.0.3 - (18/10/21)

- Resolved [#6](https://github.com/theiskaa/highlightable/issues/6) (Added case sensitive support)

#### Example:

```dart     
HighlightText(
  "Hello, Flutter!",
  highlightableWord: "flu, He",
  caseSensitive: true // Turn on case-sensitive. (as default it's false "disabled").
),
```

## 1.0.2 - (15/08/21)

- Added `detectWord` property to focus on concrete matcher words

#### Example:

```dart     
HighlightText(
  "Hello, Flutter!",
  highlightableWord: "flu, He",
  detectWords: true,
  defaultStyle: ...
  highlightStyle: ...
),
```

<img width="200" alt="stwo" src="https://user-images.githubusercontent.com/59066341/129483513-c379f0d6-d5ba-43e1-a2d7-0722aeb5dafa.png">

---

## 1.0.1 - (11/08/21)

- Fixed lower/upper case matching problem

## 1.0.0 - (11/08/21)
<p align="center">
<img width="300" src="https://user-images.githubusercontent.com/59066341/129020944-6be3379a-fc3e-4c2c-aeea-ce476fd93aae.png" alt="Package Logo">
</p>

---

```dart
HighlightText(
  'Hello World',
  highlightableWord: 'hello',
),
```

<img width="200" alt="s1" src="https://user-images.githubusercontent.com/59066341/129022549-25bd74a7-e6de-48fe-af4e-bda99106be27.png">

---

<a href="https://github.com/theiskaa/field_suggestion">
   <img src="https://img.shields.io/badge/Special%20Made%20for-FieldSuggestion-blue" alt="License: MIT"/>
</a>
<a href="https://github.com/theiskaa/highlightable-text/blob/main/LICENSE">
<img src="https://img.shields.io/badge/License-MIT-red.svg" alt="License: MIT"/>
</a>
<a href="https://github.com/theiskaa/highlightable-text/blob/main/CONTRIBUTING.md">
<img src="https://img.shields.io/badge/Contributions-Welcome-brightgreen" alt="CONTRIBUTING"/>
</a>


test/highlightable_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:highlightable/highlightable.dart';

void main() {
  late MaterialApp testApp, testApp2;
  late HighlightText highlightText, highlightText2;

  setUpAll(() {
    highlightText = HighlightText(
      'Hello Tests',
      highlight: Highlight(),
      detectWords: true,
    );

    highlightText2 = HighlightText(
      'Numbers: 20, 30, 40 will be highlighted',
      highlight: Highlight(pattern: r'\d', words: ["highlighted"]),
      detectWords: true,
      style: TextStyle(fontSize: 5),
    );

    testApp = MaterialApp(home: Scaffold(body: highlightText));
    testApp2 = MaterialApp(home: Scaffold(body: highlightText2));
  });

  group('[HighlightText]', () {
    testWidgets(
      'should expect Text widget when there is no highlighted data',
      (WidgetTester tester) async {
        await tester.pumpWidget(testApp);

        expect(find.byType(Text), findsOneWidget);
      },
    );

    testWidgets(
      'should expect generated text spans appropriate to highlight matchers',
      (WidgetTester tester) async {
        await tester.pumpWidget(testApp2);

        expect(find.byType(RichText), findsOneWidget);
      },
    );
  });
}


lib/highlightable.dart
library highlight;

import 'package:flutter/material.dart';
import 'package:highlightable/highlight_model.dart';

export 'package:highlightable/highlight_model.dart';

// Highlight mode of widget, used as private value.
// It isn't available for public API.
enum _Mode { on, off }

/// [HighlightText] is a [Text] widget alternative, that makes it easy to highlight
/// concrete words, defined from [pattern] or from [pure string].
///
/// ╭──────╮        Highlight
/// │ Data │       ╭─────────────────────────────────╮
/// ╰──────╯       │ ╭─────────╮   ╭───────────────╮ │
///    │       ╭───│ │ Pattern │ & │ Words/Letters │ │
///    │       │   │ ╰─────────╯   ╰───────────────╯ │
///    │       │   ╰─────────────────────────────────╯
///    ╰───────╯
///        │
///    ╭── ▼ ──╮         ╭─────────────────────────────────╮
///    │ Parser ───▶ ... │ Highlighted Data as Text Widget │
///    ╰───────╯         ╰─────────────────────────────────╯
///
class HighlightText extends StatelessWidget {
  /// The default string data.
  /// Like [Text] widget's first required value.
  ///
  /// For Example: When you use [Text] widget then you must to provide a string there.
  /// So, [HighlightText]'s [data] and normal [Text]'s [data] is absolutely same.
  ///
  /// Note: highlighting algorithm would search values inside that [data].
  final String data;

  /// The highlight searching model. Has two options:
  /// │─▶ [Pattern] - A regex pattern that would work for each char of [data].
  /// ╰─▶ [Words] - A list of highlightable words/letters.
  ///
  /// Example
  /// ```dart
  /// HighlightText(
  ///   highlight: Highlight(
  ///     pattern: r'\d', // Highlight each number in [data].
  ///     words: ['number', ...], // Highlight defined words in [data].
  ///   ),
  ///   ...
  /// )
  /// ```
  final Highlight highlight;

  /// The text style for the text which wouldn't be highlighted.
  ///
  /// It's the style of [data] string,
  /// Would be appended to [non-highlight-matcher] values, from [data].
  final TextStyle? style;

  /// The text style for the words/letters which would be highlighted.
  ///
  /// It's the style of [highlight] string,
  /// Would be appended to [highlight-matcher] values, from [data].
  final TextStyle highlightStyle;

  /// Enables case sensitive of parsing algorithm.
  ///
  /// If unset, defaults to the ─▶ [false].
  final bool caseSensitive;

  /// Enables only-word detection.
  /// It wouldn't highlight [matcher], if it's just a char.
  /// ╰─> {length == 1}.
  ///
  /// If unset, defaults to the ─▶ [false].
  final bool detectWords;

  HighlightText(
    this.data, {
    Key? key,
    required this.highlight,
    this.style,
    this.highlightStyle = const TextStyle(
      color: Colors.blue,
      fontWeight: FontWeight.w600,
    ),
    this.caseSensitive = false,
    this.detectWords = false,
  }) : super(key: key);

  // The highlight sequence spans.
  // Which's actual [rich-text-widget]'s childrens value.
  final List<InlineSpan> _spans = [];

  // Main method that used to generate text splans.
  // Compares [data] to [highlight] to catch highlightable and non-highlightable
  // words/letters, which would be the "NEXT" text-span's text data.
  void _parse({
    required TextStyle ds,
    required TextStyle hs,
  }) {
    if (_spans.isNotEmpty) _spans.clear();

    if (highlight.words == null && highlight.pattern == null) {
      return;
    }

    // Matching sequence representer.
    String hp = '';

    // Creates a span with [highlight-part] and [highlight-mode].
    void cutHP(_Mode mode) {
      _createSpan(hp, mode, ds: ds, hs: hs);
      hp = '';
    }

    for (var i = 0; i < data.length; i++) {
      if (hp.isEmpty) {
        hp += data[i];
        continue;
      }

      // Cut [highlight-part] as [highlight-mode-off].
      if (_isMatches(data[i]) && !_isMatches(hp)) {
        cutHP(_Mode.off);
        // Cut [highlight-part] as [highlight-mode-on].
      } else if (!_isMatches(data[i]) && _isMatches(hp)) {
        final isNotDetectable = detectWords && hp.length == 1;
        cutHP(isNotDetectable ? _Mode.off : _Mode.on);
      }

      hp += data[i];

      // Finish cutting highlight-part.
      if (hp.isNotEmpty && i == data.length - 1) {
        // Determine mode by matching of [element] and [highlight-part]+[element].
        final matches = _isMatches(data[i]) && _isMatches(hp + data[i]);
        cutHP(matches ? _Mode.on : _Mode.off);
      }
    }
  }

  // [InlineSpan] generator.
  // Generates a text span appropriate by given data and mode.
  //
  // Mode determines the style of generated text-span.
  // If mode is on, text-span's style would be [highlightStyle].
  // If not, style would be (default)[style].
  void _createSpan(
    String data,
    _Mode mode, {
    required TextStyle ds,
    required TextStyle hs,
  }) {
    final textSpan = TextSpan(
      text: data,
      style: (mode == _Mode.on) ? hs : ds,
    );

    _spans.add(textSpan);
  }

  // Matching checker for [parse] method.
  // Would be used to check matching of one letter or split word.
  bool _isMatches(String data) {
    // Ignore case sensitive.
    if (!caseSensitive) data = data.toLowerCase();

    // Check matching by pattern.
    if (highlight.pattern != null &&
        RegExp(highlight.pattern!).hasMatch(data)) {
      return true;
    }

    // Check matching by words/words.
    if (highlight.words != null && highlight.words!.isNotEmpty) {
      for (var i = 0; i < highlight.words!.length; i++) {
        var word = highlight.words![i];

        // Ignore case sensitive.
        if (!caseSensitive) word = word.toLowerCase();
        if (word.contains(data) || data.contains(word)) return true;
      }
    }

    return false;
  }

  @override
  Widget build(BuildContext context) {
    // Define style of normal text.
    final DefaultTextStyle defaultTextStyle = DefaultTextStyle.of(context);
    TextStyle effectiveStyle = style ?? defaultTextStyle.style;
    if (style == null || style!.inherit) {
      effectiveStyle = defaultTextStyle.style.merge(style);
    }

    // Generate highlight-sequence text-spans.
    _parse(ds: effectiveStyle, hs: highlightStyle);

    // Build the default text, if there is no spans.
    if (_spans.isEmpty) return Text(data, style: effectiveStyle);

    // Build the highlighted text with rich-text.
    return RichText(text: TextSpan(text: '', children: _spans));
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:highlightable/highlightable.dart';

void main() => runApp(const App());

class App extends StatelessWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => const MaterialApp(home: Home());
}

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Highlightable Examples")),
      body: Center(
        child: Padding(
          padding: const EdgeInsets.all(15),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              // Basic usage.
              HighlightText(
                'Only numbers: [1, 25, 50, ...] will be highlighted',
                // would highlight only numbers.
                highlight: const Highlight(pattern: r'\d'),
              ),

              const SizedBox(height: 50),

              // Custom Usage
              HighlightText(
                "Hello, Flutter!",
                // Would highlight only "Flutter"
                // full word 'cause [detectWords] is enabled.
                highlight: const Highlight(
                  words: ["Flutter"],
                ),
                caseSensitive: true, // Turn on case-sensitive.
                detectWords: true,
                style: const TextStyle(
                  fontSize: 25,
                  color: Colors.black,
                  fontWeight: FontWeight.bold,
                ),
                highlightStyle: const TextStyle(
                  fontSize: 25,
                  letterSpacing: 2.5,
                  color: Colors.white,
                  backgroundColor: Colors.blue,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
